x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts

```
@@ -179,14 +189,19 @@ export class StreamsClient {
         definition,
         logger,
         scopedClusterClient,
+        unwiredRoot: await this.getUnwiredRootForStream(definition),
         isServerless: this.dependencies.isServerless,
       });
     } else if (isIngestStream(definition)) {
+      const dataStream = await this.getDataStream(definition.name);
+      if (!dataStream) {
+        throw new MalformedStream(`Data stream ${definition.name} could not be found`);
```

this should already throw a 404 if it's not found, right?

---

That's mostly to make the type checker happy - better than the `!` operator and a weird arcane error a couple months and refactorings down the line.

---

yeah but shouldn't getDataStream itself throw a 404 if it isn't found?

---

No, in some paths this is a valid case we should handle - if for example a user adds a dashboard and some processing to a classic stream and then the underlying data stream is deleted, it should still show and indicate that it's "orphaned". This is not fully implemented today, but I will complete it in a follow-up.

In this particular place as you mentioned we can't really complete the request, so it throws here.

---

What i mean is, any GET call with a specific id imho should fail with a 404, and not a 200 and undefined - which is to say, i think getDataStream() should return the data stream or return a 404 if not found, and then handling that error is up to the consumer. If we allow it to return undefined we end up pushing the assertion to the user which i think we will want to avoid

---

> any GET call with a specific id imho should fail with a 404, and not a 200 and undefined

For sure, on the API level that is how it works. but the internal `getDataStream` method is one level below it and has different semantics.

> If we allow it to return undefined we end up pushing the assertion to the user which i think we will want to avoid

Why do you think so? I think it's a sensible API with clear semantics and we enforce via types that the consumer does the right thing. The alternative would be to have a throwing `getDataStream` and an `existsDataStream`, but that would imply two roundtrips we can avoid here and I think in most places we want to handle the situation that a data stream we thought would exist in a graceful way instead of throwing. Like here: https://github.com/elastic/kibana/blob/8d4a70c5e5a0ad9fd985d71d2f107bc29b330a92/x-pack/solutions/observability/plugins/streams/server/routes/streams/crud/read_stream.ts#L49